### PR TITLE
Feature 111 ground segmentation

### PR DESCRIPTION
**Pull request by Jebediah**
Migration of ground segmentation class from previous @arunxdas's old project into libwave. See #111 .

Cleaned up code a bit, added a little functionality in terms of configuration, and added some doxygen.

### CODE REVIEW

File: wave_matching/CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,8 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 2.8.3)
 PROJECT(wave_matching)
 
-SET(CMAKE_BUILD_TYPE RELEASE)
+SET(CMAKE_CXX_FLAGS "-std=gnu++11 ${CMAKE_CXX_FLAGS}")
```
Comment (leokoppel) [inline wave_matching/CMakeLists.txt]: `c++11`

File: wave_matching/CMakeLists.txt [hunk 2]
```diff
@@ -1,7 +1,8 @@
...
+SET(CMAKE_BUILD_TYPE release)
```
Comment (leokoppel) [inline wave_matching/CMakeLists.txt]: Do not overwrite BUILD_TYPE if already set. Example:
https://github.com/wavelab/okvis_ros/commit/89ab4be8a6c1d956fa7bccd659f829d547891fc9

File: wave_matching/config/groundSegmentation.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+# GP Kernel hyperparameters
+gp_lengthparameter:       10
+gp_covariancescale:       1
+gp_modelnoise:            0.3
+gp_groundmodelconfidence: 5
+gp_grounddataconfidence:  5
+
+gp_groundthreshold:       0.3 #Ground Height Threshold
+robotheight:              1.2 #Used for drivable surface detections I think
+seeding_maxrange:         50  #Maximum range for seed points
```
Comment (leokoppel) [inline wave_matching/config/groundSegmentation.yaml]: What is a "seed point"? Is there additional documentation? At least a reference to a paper that defines these terms?

File: wave_matching/include/wave/matching/PointcloudXYZGD.hpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
+
+/************************************************************************
+ *
+ *
+ *  Copyright 2015  Arun Das (University of Waterloo)
```
Comment (leokoppel) [inline wave_matching/include/wave/matching/PointcloudXYZGD.hpp]: Change to `@authors` command?

File: wave_matching/include/wave/matching/groundSegmentationParams.hpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    // seeding parameters
+    double max_seed_range;   // meters
+    double max_seed_height;  // meters
+
+    int num_bins_a;
+    int num_bins_l;
+
+    // set default parameters
+    GroundSegmentationParams() {
+        rmax = 100;
```
Comment (leokoppel) [inline wave_matching/include/wave/matching/groundSegmentationParams.hpp]: Move these assignments to default member initializers ([#2 here](http://en.cppreference.com/w/cpp/language/data_members#Member_initialization))

File: wave_matching/src/groundSegmentation.cpp [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+                obsSum += Vec3(curCell->binPoints[j].x,
+                               curCell->binPoints[j].y,
+                               curCell->binPoints[j].z);
+                numObs++;
+            }
+            // mean of obs points
+            curCell->obsMean = obsSum / numObs;
+        }
+    } else {
+        LOG_INFO("WARNING:Insufficnent Model for angular slice");
```
Comment (leokoppel) [inline wave_matching/src/groundSegmentation.cpp]: typo

File: wave_matching/config/groundSegmentation.yaml (hunk 3 above)
Comment (Jebediah) [reply wave_matching/config/groundSegmentation.yaml]: I think it is following the approach in https://link.springer.com/article/10.1007/s10846-013-9889-4. I think @arunxdas 's thesis also mentions it.

File: wave_matching/CMakeLists.txt (hunk 1 above)
Comment (Jebediah) [reply wave_matching/CMakeLists.txt]: do you mean change gnu++11 to c++11?

Comment (leokoppel) [reply wave_matching/CMakeLists.txt]: Yes. Actually it is already set in top-level CMakeLists.txt so can remove

File: wave_matching/config/groundSegmentation.yaml (hunk 3 above)
Comment (leokoppel) [reply wave_matching/config/groundSegmentation.yaml]: Thanks, add to code somewhere

Comment (Jebediah) [reply wave_matching/config/groundSegmentation.yaml]: Let's wait until Arun confirms that it's the right one.

File: wave_matching/include/wave/matching/PointcloudXYZGD.hpp (hunk 4 above)
Comment (Jebediah) [reply wave_matching/include/wave/matching/PointcloudXYZGD.hpp]: nah, they did write it. I just doubt that they actually retain copyright.

Comment (leokoppel) [reply wave_matching/include/wave/matching/PointcloudXYZGD.hpp]: I mean you can replace this header with one consistent with all our other files - Doxygen formatting, `@file` command, etc. Then in the same place, there can be `@authors` listed. The names and university and email addresses can appear there.

**Pull request merged**